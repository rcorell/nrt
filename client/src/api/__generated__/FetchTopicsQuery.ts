/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: FetchTopicsQuery
// ====================================================

export interface FetchTopicsQuery_topics {
  __typename: "Topic";
  title: string;
  description: string;
}

export interface FetchTopicsQuery {
  topics: FetchTopicsQuery_topics[];
}
