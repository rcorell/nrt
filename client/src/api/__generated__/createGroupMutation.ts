/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: CreateGroupMutation
// ====================================================

export interface CreateGroupMutation_createGroup {
    __typename: 'Topic';
    id: string;
    createdAt: any;
}

export interface CreateGroupMutation {
    createGroup: CreateGroupMutation_createGroup;
}

export interface CreateGroupMutationVariables {
    name: string;
    description?: string | null;
}
