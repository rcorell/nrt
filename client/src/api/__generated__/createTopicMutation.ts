/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: CreateTopicMutation
// ====================================================

export interface CreateTopicMutation_createTopic {
    __typename: 'Topic';
    id: string;
    createdAt: any;
}

export interface CreateTopicMutation {
    createTopic: CreateTopicMutation_createTopic;
}

export interface CreateTopicMutationVariables {
    title: string;
    description?: string | null;
}
