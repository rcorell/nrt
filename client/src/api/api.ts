import { ApolloClient, ApolloLink, gql, HttpLink, InMemoryCache } from '@apollo/client';

const AUTHENTICATION_TOKEN_NAME = 'token';

const httpLink = new HttpLink({ uri: 'http://localhost:4000' });

const authLink = new ApolloLink((operation, forward) => {
    const token = localStorage.getItem(AUTHENTICATION_TOKEN_NAME);

    operation.setContext({
        headers: {
            authorization: token ? `Bearer ${token}` : ''
        }
    });

    return forward(operation);
});

export const apolloClient = new ApolloClient({
    link: authLink.concat(httpLink),
    cache: new InMemoryCache()
});

export const isAuthenticated = () => {
    return Boolean(localStorage.getItem(AUTHENTICATION_TOKEN_NAME));
};

export const setAuthenticationToken = (token: string) => {
    localStorage.setItem(AUTHENTICATION_TOKEN_NAME, token);
};

export const fetchTopicsQueryString = gql`
    query FetchTopicsQuery {
        topics {
            title
            description
        }
    }
`;

export const CreateGroupMutation = gql`
    mutation CreateGroupMutation($name: String!, $description: String) {
        createGroup(name: $name, description: $description) {
            id
            createdAt
        }
    }
`;

export const SignupMutation = gql`
    mutation SignupMutation($email: String!, $name: String!, $password: String!) {
        signup(email: $email, name: $name, password: $password) {
            token
        }
    }
`;

export const loginMutationString = gql`
    mutation LoginMutation($email: String!, $password: String!) {
        login(email: $email, password: $password) {
            token
        }
    }
`;

export const CreateTopicMutation = gql`
    mutation CreateTopicMutation($title: String!, $description: String) {
        createTopic(title: $title, description: $description) {
            id
            createdAt
        }
    }
`;
