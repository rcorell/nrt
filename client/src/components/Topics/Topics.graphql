query FetchTopics {
        topics {
            id
            title
            description
            group {
                id
                name
            }
        }
    }