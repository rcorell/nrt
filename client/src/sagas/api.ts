import { SagaIterator } from 'redux-saga';
import { all, takeLatest } from 'redux-saga/effects';
import { ActionType } from 'typesafe-actions';

import * as signupActions from '../redux/signup/actions';

export function* watcher(): SagaIterator<void> {
    // yield call(_checkAuthenticated);

    yield all([
        takeLatest(signupActions.signup, _signup),
    ]);
}

export function* _signup(action: ActionType<typeof signupActions.signup>): SagaIterator<void> {
	const { email, name, password } = action.payload;

	console.log(email, name, password);
}
