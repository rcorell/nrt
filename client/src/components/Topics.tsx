import { useQuery } from '@apollo/client';
import _ from 'lodash';
import MaterialTable from 'material-table';
import React from 'react';

import { FetchTopicsQuery } from 'src/api/__generated__/FetchTopicsQuery';
import { fetchTopicsQuery } from 'src/api/api';
import { LOADING_TEXT } from 'src/components/shared';
import { setBrowserTitle } from 'src/utils';

const COLUMN_DEFINITIONS = [
    { field: 'title', title: 'Title' },
    { field: 'description', title: 'Description' },
    { field: 'group', title: 'Group' }
];

export const Topics: React.FC = () => {
    setBrowserTitle('Topics');

    const { data, error, loading } = useQuery<FetchTopicsQuery>(fetchTopicsQuery, { fetchPolicy: 'network-only' });

    if (loading) {
        return <div>{LOADING_TEXT}</div>;
    }

    if (error) {
        return <div>Error: {JSON.stringify(error)}</div>;
    }

    const topics = _.cloneDeep(data!.topics).map((topic) => ({ ...topic, group: topic.group.name }));

    return (
        <MaterialTable
            title="Topics"
            columns={COLUMN_DEFINITIONS}
            data={topics}
            options={{
                pageSize: 20,
                pageSizeOptions: [20, 50, 100]
            }}
        />
    );
};
