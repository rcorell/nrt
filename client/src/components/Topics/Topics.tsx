import _ from 'lodash';
import MaterialTable from 'material-table';
import React from 'react';

import { LOADING_TEXT } from 'src/components/shared';
import { setBrowserTitle } from 'src/utils';
import { useTopics } from './Topics.hook';

const COLUMN_DEFINITIONS = [
    { field: 'title', title: 'Title' },
    { field: 'description', title: 'Description' },
    { field: 'group', title: 'Group' }
];

export const Topics: React.FC = () => {
    setBrowserTitle('Topics');

    const useTopicsHook = useTopics();

    if (useTopicsHook.loading) {
        return <div>{LOADING_TEXT}</div>;
    }

    if (useTopicsHook.error) {
        return <div>Error: {JSON.stringify(useTopicsHook.error)}</div>;
    }

    const topics = _.cloneDeep(useTopicsHook.topics!).map((topic) => ({ ...topic, group: topic.group.name }));

    return (
        <MaterialTable
            title="Topics"
            columns={COLUMN_DEFINITIONS}
            data={topics}
            options={{
                pageSize: 20,
                pageSizeOptions: [20, 50, 100]
            }}
        />
    );
};
