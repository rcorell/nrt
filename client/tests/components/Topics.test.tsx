import { screen, waitForElementToBeRemoved } from '@testing-library/react';

import { FetchTopicsQuery } from 'src/api/__generated__/FetchTopicsQuery';
import { fetchTopicsQueryString } from 'src/api/api';
import { Topics } from 'src/components/Topics';

import { renderComponent } from 'tests/testHelpers';

describe('Topics', () => {
    const LOADING_TEXT = 'Loading...';

    describe('snapshots', () => {
        it('loading', () => {
            expect(renderComponent(Topics)).toMatchSnapshot();
        });

        it('success', async () => {
            const data: FetchTopicsQuery = {
                topics: [{ __typename: 'Topic', description: 'test-description', title: 'test-title' }]
            };

            renderComponent(Topics, fetchTopicsQueryString, {}, { data });

            await waitForElementToBeRemoved(() => screen.getByText(LOADING_TEXT));

            // Material UI devs literally hate snapshots
            expect(screen.queryByText('test-description')).toBeInTheDocument();
            expect(screen.queryByText('test-title')).toBeInTheDocument();
        });

        it('network error', async () => {
            const networkError = new Error('Network-error');

            const topicsRender = renderComponent(Topics, fetchTopicsQueryString, {}, { error: networkError });

            await waitForElementToBeRemoved(() => screen.getByText(LOADING_TEXT));

            expect(screen.queryByText('No results')).toBeNull();

            expect(topicsRender).toMatchSnapshot();
        });

        it('Malformed results', async () => {
            const topicsRender = renderComponent(Topics, fetchTopicsQueryString, {}, { data: null });

            await waitForElementToBeRemoved(() => screen.getByText(LOADING_TEXT));

            expect(topicsRender).toMatchSnapshot();
        });
    });
});
