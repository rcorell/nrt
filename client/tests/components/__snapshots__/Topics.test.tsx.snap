// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Topics snapshots Malformed results 1`] = `
<div>
  <div>
    Error: 
    {"graphQLErrors":[],"networkError":{},"message":"Mocked response should contain either result or error: {\\"query\\":\\"query FetchTopicsQuery {\\\\n  topics {\\\\n    id\\\\n    title\\\\n    description\\\\n  }\\\\n}\\\\n\\"}"}
  </div>
</div>
`;

exports[`Topics snapshots loading 1`] = `
<div>
  <div>
    Loading...
  </div>
</div>
`;

exports[`Topics snapshots network error 1`] = `
<div>
  <div>
    Error: 
    {"graphQLErrors":[],"networkError":{},"message":"Network-error"}
  </div>
</div>
`;
